OAK-11486: Add failing test case for session-mapped non-registered

kwin opened the pull request:
namespace URIs

kwin committed:
OAK-11486: Add failing test case for session-mapped non-registered
namespace URIs

mbaedke committed:
OAK-11486: Session-local namespace mappings for unregistered namespace URIs lead to unexpected ISE for Node.getName() or Node.getPath()

LocalNameMapper now checks if only a local namespace mapping exists.

kwin commented on oak-core/src/main/java/org/apache/jackrabbit/oak/namepath/impl/LocalNameMapper.java:
```diff
@@ -64,8 +64,8 @@ public synchronized String getJcrName(@NotNull String oakName) {
                 if (uri == null) {
-                    throw new IllegalStateException(
-                            "No namespace mapping found for " + oakName);
+                    //the namespace has a local mapping, but no global mapping, which is fine
```
I don't think this comment is accurate, this only means that there is some local namespace mapping (irrespective for which prefix/uri combination) but the prefix is not registered globally.

mbaedke replied:
Yes, that is true. I'll fix it.

mbaedke committed:
OAK-11486: Session-local namespace mappings for unregistered namespace URIs lead to unexpected ISE for Node.getName() or Node.getPath()

Corrected faulty fix attempt.

kwin commented on oak-core/src/main/java/org/apache/jackrabbit/oak/namepath/impl/LocalNameMapper.java:
```diff
@@ -91,9 +91,10 @@ public synchronized String getJcrName(@NotNull String oakName) {
                     }
                 }
+            } else {
+                throw new IllegalStateException("No namespace mapping found for " + oakName);
```
With that we are back at the same exception as before :-)

kwin replied:
actually we are now throwing an exception even earlier (i.e. for the first getName() without any mapping)

mbaedke replied:
That was an intermediate version. This current one should hopefully do, except for the problem I mentioned on Jira, which probably is something completely different.

kwin replied:
I wouldn't understand why `Session.addNode("foo:bar").getName()` wouldn't end up here, because

a) uri is null (there is no namespace URI registered for "foo" neither globally nor locally
b) there is no local namespace mapping for "foo"

This previously didn't lead to an exception. Now it does.

mbaedke committed:
OAK-11486: Session-local namespace mappings for unregistered namespace URIs lead to unexpected ISE for Node.getName() or Node.getPath()

Fixed handling of local namespace mappings with globally unknown URIs; code cleanup.

kwin commented:
@mbaedke Do you plan to pick this up again?
